### DISCUSSION TITLE
How to post a command in Rabbit MQ that uses a Header Exchange with parameters?

### DISCUSSION THREAD

**Discussion by leonErhard** [BrighterCommand/Brighter#2491] [General]
I was reading the docs about how to configure and post commands with External Brokers like RabbitMQ, but I notice there's no content about  how to configure and send messages in an Header Exchange that has parameters to use the correct queue. How can I Post a command in a Header Exchange with parameters ?

**Comment by DevJonny** [comment]
Hi @leonErhard.

You can [configure Brighter to use the Headers exchange](https://brightercommand.gitbook.io/paramore-brighter-documentation/brighter-configuration/rabbitmqconfiguration#connection) when you setup the `RmqMessagingGatewayConnection`, so using the example in the [RMQ configuration section of docs](https://brightercommand.gitbook.io/paramore-brighter-documentation/brighter-configuration/rabbitmqconfiguration#connection). So when you create the `Exchange` object you can pass `ExchangeType.Headers` in the `type` parameter.

```csharp
public void ConfigureServices(IServiceCollection services)
{
    services.AddBrighter(...)
        .UseExternalBus(new RmqProducerRegistryFactory(
                    new RmqMessagingGatewayConnection
                    {
                        Name = "MyCommandConnection",
                        AmpqUri = new AmqpUriSpecification(
                            new Uri("amqp://guest:guest@localhost:5672")
                            connectionRetryCount: 5,
                            retryWaitInMilliseconds: 250,
                            circuitBreakerTimeInMilliseconds = 30000
                        ),
                        Exchange = new Exchange("paramore.brighter.exchange", type: ExchangeType.Headers, durable: true, supportDelay: true),
                        DeadLetterExchange = new Exchange("paramore.brighter.exchange.dlq", durable: true, supportDelay: false),
                        Heartbeat = 15,
                        PersistMessages = true
                    },
            ... //publication, see below
        ).Create()
}
```

Hope this helps!

**Reply by leonErhard** [reply]
Thanks! How can i Post a command with a header parameter "mail":true ?

**Reply by DevJonny** [reply]
Sorry for the slow reply @leonErhard it completely slipped my mind. Adding a header to the RMQ message can be done when creating your [Message Mappers](https://brightercommand.gitbook.io/paramore-brighter-documentation/brighter-request-handlers-and-middleware-pipelines/messagemappers#writing-a-message-mapper). All you need to do, besides the aforementioned code, is add something along the lines of 

```
...
header.Bag.TryAdd("mail", true);
...
```

So when Brighter serialises your Command/Event into a RMQ message, all the items in the `Bag` dictionary of the message are [added as headers to the message](https://github.com/BrighterCommand/Brighter/blob/c79f8cef886fd3550c3da9c201fcebbb420d67da/src/Paramore.Brighter.MessagingGateway.RMQ/RmqMessageCreator.cs#L83).
